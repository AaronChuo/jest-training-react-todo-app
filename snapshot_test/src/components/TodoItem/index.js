import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import TodoTextInput from 'components/TodoTextInput';
import style from 'todomvc-app-css/index.css';

export default class TodoItem extends Component {
  static propTypes = {
    todo: PropTypes.object.isRequired,
    editTodo: PropTypes.func.isRequired,
    deleteTodo: PropTypes.func.isRequired,
    completeTodo: PropTypes.func.isRequired,
  };

  state = {
    editing: false,
  };

  handleDoubleClick = () => {
    this.setState({ editing: true });
  };

  handleSave = (id, text) => {
    if (text.length === 0) {
      this.props.deleteTodo({ id });
    } else {
      this.props.editTodo({ id, text });
    }
    this.setState({ editing: false });
  };

  render () {
    const { todo, completeTodo, deleteTodo } = this.props;

    let element;
    if (this.state.editing) {
      element = (
        <TodoTextInput
          text={todo.text}
          editing={this.state.editing}
          onSave={text => this.handleSave(todo.id, text)}
        />
      );
    } else {
      element = (
        <div className={style['view']}>
          <input
            className={style['toggle']}
            type="checkbox"
            checked={todo.completed}
            onChange={() => completeTodo({ id: todo.id })}
          />
          <label onDoubleClick={this.handleDoubleClick}>
            {todo.text}
          </label>
          <button className={style['destroy']} onClick={() => deleteTodo({ id: todo.id })} />
        </div>
      );
    }

    return (
      <li
        className={classnames({
          [style['completed']]: todo.completed,
          [style['editing']]: this.state.editing,
        })}
      >
        {element}
      </li>
    );
  }
}
