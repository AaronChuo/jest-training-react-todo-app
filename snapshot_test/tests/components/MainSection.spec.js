import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import MainSection from 'components/MainSection';
import TodoItem from 'components/TodoItem';
import Footer from 'components/Footer';
import { todoList } from 'mock-data';
import {
  SHOW_ALL,
  SHOW_COMPLETED,
  SHOW_ACTIVE,
} from 'constants/TodoFilters';

describe('<MainSection />', () => {
  let actions;
  let wrapper;

  beforeEach(() => {
    actions = {
      clearCompleted: jest.fn(),
      completeAll: jest.fn(),
      editTodo: jest.fn(),
      deleteTodo: jest.fn(),
      completeTodo: jest.fn(),
    };
    wrapper = shallow(
      <MainSection
        todos={todoList}
        actions={actions}
      />
    );
  });

  it('should render initial component', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render component with filter SHOW_ACTIVE', () => {
    wrapper.setState({ filter: SHOW_ACTIVE });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render component with filter SHOW_COMPLETED', () => {
    wrapper.setState({ filter: SHOW_COMPLETED });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates show todo with filter', () => {
    wrapper.find(Footer).simulate('show', SHOW_ACTIVE);
    expect(wrapper.state()).toMatchSnapshot();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates clear completed todo', () => {
    wrapper.find(Footer).simulate('clearCompleted');
    expect(actions.clearCompleted).toHaveBeenCalled();
  });
});
