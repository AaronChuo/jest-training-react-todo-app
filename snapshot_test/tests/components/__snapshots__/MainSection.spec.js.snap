// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainSection /> should render component with filter SHOW_ACTIVE 1`] = `
<section
  className="main"
>
  <input
    checked={false}
    className="toggle-all"
    onChange={[Function]}
    type="checkbox"
  />
  <ul
    className="todo-list"
  >
    <TodoItem
      clearCompleted={[Function]}
      completeAll={[Function]}
      completeTodo={[Function]}
      deleteTodo={[Function]}
      editTodo={[Function]}
      todo={
        Object {
          "completed": false,
          "id": 0,
          "text": "Use Redux",
        }
      }
    />
  </ul>
  <Footer
    activeCount={1}
    completedCount={1}
    filter="show_active"
    onClearCompleted={[Function]}
    onShow={[Function]}
  />
</section>
`;

exports[`<MainSection /> should render component with filter SHOW_COMPLETED 1`] = `
<section
  className="main"
>
  <input
    checked={false}
    className="toggle-all"
    onChange={[Function]}
    type="checkbox"
  />
  <ul
    className="todo-list"
  >
    <TodoItem
      clearCompleted={[Function]}
      completeAll={[Function]}
      completeTodo={[Function]}
      deleteTodo={[Function]}
      editTodo={[Function]}
      todo={
        Object {
          "completed": true,
          "id": 1,
          "text": "Use CodeSandbox",
        }
      }
    />
  </ul>
  <Footer
    activeCount={1}
    completedCount={1}
    filter="show_completed"
    onClearCompleted={[Function]}
    onShow={[Function]}
  />
</section>
`;

exports[`<MainSection /> should render initial component 1`] = `
<section
  className="main"
>
  <input
    checked={false}
    className="toggle-all"
    onChange={[Function]}
    type="checkbox"
  />
  <ul
    className="todo-list"
  >
    <TodoItem
      clearCompleted={[Function]}
      completeAll={[Function]}
      completeTodo={[Function]}
      deleteTodo={[Function]}
      editTodo={[Function]}
      todo={
        Object {
          "completed": false,
          "id": 0,
          "text": "Use Redux",
        }
      }
    />
    <TodoItem
      clearCompleted={[Function]}
      completeAll={[Function]}
      completeTodo={[Function]}
      deleteTodo={[Function]}
      editTodo={[Function]}
      todo={
        Object {
          "completed": true,
          "id": 1,
          "text": "Use CodeSandbox",
        }
      }
    />
  </ul>
  <Footer
    activeCount={1}
    completedCount={1}
    filter="show_all"
    onClearCompleted={[Function]}
    onShow={[Function]}
  />
</section>
`;

exports[`<MainSection /> simulates show todo with filter 1`] = `
Object {
  "filter": "show_active",
}
`;

exports[`<MainSection /> simulates show todo with filter 2`] = `
<section
  className="main"
>
  <input
    checked={false}
    className="toggle-all"
    onChange={[Function]}
    type="checkbox"
  />
  <ul
    className="todo-list"
  >
    <TodoItem
      clearCompleted={[Function]}
      completeAll={[Function]}
      completeTodo={[Function]}
      deleteTodo={[Function]}
      editTodo={[Function]}
      todo={
        Object {
          "completed": false,
          "id": 0,
          "text": "Use Redux",
        }
      }
    />
  </ul>
  <Footer
    activeCount={1}
    completedCount={1}
    filter="show_active"
    onClearCompleted={[Function]}
    onShow={[Function]}
  />
</section>
`;
